src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -245,12 +246,103 @@ private BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOpe
             bool rightNullable = right.Type?.IsNullableType() == true;
             bool isNullable = leftNullable || rightNullable;
 
-            TypeSymbol leftTupleType = MakeConvertedType(operators.SelectAsArray(o => o.LeftConvertedTypeOpt), node.Left, leftParts, isNullable, compilation, diagnostics);
-            TypeSymbol rightTupleType = MakeConvertedType(operators.SelectAsArray(o => o.RightConvertedTypeOpt), node.Right, rightParts, isNullable, compilation, diagnostics);
+            TypeSymbol leftTupleType = MakeConvertedType(operators.SelectAsArray(o => o.LeftConvertedTypeOpt), node.Left, leftParts, leftNames, isNullable, compilation, diagnostics);
+            TypeSymbol rightTupleType = MakeConvertedType(operators.SelectAsArray(o => o.RightConvertedTypeOpt), node.Right, rightParts, rightNames, isNullable, compilation, diagnostics);
 
             return new TupleBinaryOperatorInfo.Multiple(operators, leftTupleType, rightTupleType);
         }
 
+        /// <summary>
+        /// If an element in a tuple literal has an explicit name which doesn't match the name on the other side, we'll warn.
+        /// The user can either remove the name, or fix it.
+        ///
+        /// This method handles two expressions, each of which is either a tuple literal or an expression with tuple type.
+        /// In a tuple literal, each element can have an explicit name, an inferred name or no name.
+        /// In an expression of tuple type, each element can have a name or not.
+        /// </summary>
+        private static void ReportNamesMismatchesIfAny(BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            bool leftIsTupleLiteral = left.Kind == BoundKind.TupleLiteral;
+            bool rightIsTupleLiteral = right.Kind == BoundKind.TupleLiteral;
+
+            if (!leftIsTupleLiteral && !rightIsTupleLiteral)
+            {
+                return;
+            }
+
+            var (leftNames, leftInferred, leftLocations) = getDetails(left);
```

Comment:
>var [](start = 12, length = 3)

Please spell out types

Reply:
I'd much rather not in this case: `(ImmutableArray<string> leftNames, ImmutableArray<bool> leftInferred, ImmutableArray<BoundExpression> leftLocations)` is really long :-(

---
In reply to: [173939157](https://github.com/dotnet/roslyn/pull/25396#discussion_r173939157) [](ancestors = 173939157)

Reply:
> I'd much rather not in this case: (ImmutableArray<string> leftNames, ImmutableArray<bool> leftInferred, ImmutableArray<BoundExpression> leftLocations) is really long :-(

Sorry, there is a rule and we should follow it. The code is not readable without the types.  

---
In reply to: [173944067](https://github.com/dotnet/roslyn/pull/25396#discussion_r173944067) [](ancestors = 173944067,173939157)